azure/azure.go

```diff
@@ -46,11 +58,112 @@ var (
+func NewAzureClient() (cloudops.Ops, error) {
+
+	if onAzure() {
+		logrus.Info("Running on Azure IaaS VM")
+
+		client := &http.Client{}
+
+		req, _ := http.NewRequest("GET", metadataAPIEndpoint, nil)
+		req.Header.Add("Metadata", "True")
+		q := req.URL.Query()
+		q.Add("format", "json")
+		q.Add("api-version", metadataAPIVersion)
+		req.URL.RawQuery = q.Encode()
+
+		resp, err := client.Do(req)
+		if err != nil {
+			logrus.Errorf("Error occured while getting instance metadata from Azure Metadata API. Error:[%v]", err)
+			return nil, err
+		}
+
+		defer resp.Body.Close()
+		if resp.StatusCode != 200 {
+			return nil, fmt.Errorf("Error querying Azure metadata: Code %d returned for url %s", resp.StatusCode, req.URL)
+		}
+		respBody, _ := ioutil.ReadAll(resp.Body)
+
+		if len(respBody) == 0 {
+			return nil, fmt.Errorf("Error querying Azure metadata: Empty response")
+		}
+
+		metadata := make(map[string]interface{})
+		err = json.Unmarshal(respBody, &metadata)
+		if err != nil {
+			return nil, fmt.Errorf("Error parsing Azure metadata: %v", err)
+		}
+
+		var vmID, scalesetName, subscriptionID, resourceGroupName, managedClusterName, agentPoolName string
+		var scalesetNametemp, vmIDtemp, subscriptionIDtemp interface{}
+		var exists bool
+
+		if resourceGroup, ok := metadata[resourceGroupNameKey]; ok {
+			temp := resourceGroup.(string)
+			if len(strings.Split(temp, "_")) == 2 {
+				resourceGroupName = strings.Split(temp, "_")[1]
+				managedClusterName = strings.Split(temp, "_")[2]
+			}
+		}
+
+		if scalesetNametemp, exists = metadata[scaleSetNameKey]; exists {
+			scalesetName := scalesetNametemp.(string)
+			agentPoolName = strings.Split(scalesetName, "-")[1]
+		}
+
+		if vmIDtemp, exists = metadata[vmIDKey]; exists {
+			vmID = vmIDtemp.(string)
+		}
+
+		if subscriptionIDtemp, exists = metadata[subscriptionIDKey]; exists {
+			subscriptionID = subscriptionIDtemp.(string)
+		}
+
+		return NewClientWithPoolName(vmID, scalesetName, subscriptionID, resourceGroupName, managedClusterName, agentPoolName)
+
+	}
+	logrus.Info("Not running on Azure IaaS VM")
+	return NewEnvClient()
+}
+
+// onAzure checks if the process is running inside Azure VM
+// by check rechability of Metadata API service.
+func onAzure() bool {
+
+	client := &http.Client{}
+	req, _ := http.NewRequest("GET", metadataAPIEndpoint, nil)
+	req.Header.Add("Metadata", "True")
+	q := req.URL.Query()
+	q.Add("format", "json")
+	q.Add("api-version", metadataAPIVersion)
+	req.URL.RawQuery = q.Encode()
+
+	resp, err := client.Do(req)
+	if err != nil {
+		logrus.Errorf("Error occured while getting instance metadata from Azure Metadata API. Error:[%v]."+
+			"Probably not running on Azure IaaS VMs created by Azure Resource Manager.", err)
+		return false
+	}
+
+	defer resp.Body.Close()
+	if resp.StatusCode != 200 {
+		logrus.Errorf("Error querying Azure metadata: Code %d returned for url %s", resp.StatusCode, req.URL)
+		return false
+	}
```

Comment:
duplicate of snipped in `NewAzureClient()`. Put in common method and have it satisfy both use cases